File: mac/Sources/DestruidorDeRed/CooldownModel.swift
import Foundation

// Modelo do sistema de Cooldown/Alarme (#33).
// - CATÁLOGO: templates lidos de `data/cooldowns.json` (batalhas, tiers de berry, 64 berries).
// - ESTADO: o que o usuário criou/marcou (personagens + marcações), salvo em UserDefaults.
//
// Regra de robustez: a VERDADE é sempre um timestamp absoluto (epoch em MILISSEGUNDOS, igual ao
// protótipo/sync). Contador e alarme são derivados; reconciliar no startup a partir dos timestamps.

/// Milissegundos desde 1970 (mesma base do protótipo web e do futuro sync).
func nowMs() -> Double { Date().timeIntervalSince1970 * 1000 }

/// Nome bilíngue (PT/EN) vindo do catálogo.
struct LocalizedName: Codable, Equatable {
    var pt: String
    var en: String
    var localized: String { AppLang.current == .en ? en : pt }
}

// MARK: - Catálogo (data/cooldowns.json)

/// Tarefa de batalha com cooldown simples (um timer).
struct BattleTask: Codable, Identifiable, Equatable {
    var id: String
    var category: String
    var name: LocalizedName
    var hours: Double
    var color: String
    var confidence: String
    var defaultOn: Bool
    /// Especificação do ícone: "region:x" | "trainer:x" | "item:x" | "sprite:x" | "sf:símbolo".
    /// Opcional (nil = cai no ponto colorido). (campos extras type/note/source são ignorados.)
    var icon: String?
    /// Agrupador na UI (ex.: "elite4" junta as 5 ligas num submenu). nil = tarefa avulsa.
    var group: String?
}

/// Tier de berry: guarda a matemática do ciclo (uma vez), compartilhada por várias berries.
struct BerryTier: Codable, Identifiable, Equatable {
    var tier: String
    var growthHours: Double        // plantar -> pronta pra colher (fixo)
    var wiltHours: Double          // janela de colheita depois de pronta
    var waterWindowsHours: [Double] // limite (h após plantar) de cada rega
    var waterLeadHours: Double      // disparar o lembrete tantas h ANTES do limite (adiantar é seguro)
    var yield: String
    var id: String { tier }
}

/// Uma berry específica (nome/ícone) que aponta pro seu tier.
struct BerryDef: Codable, Identifiable, Equatable {
    var id: String
    var tier: String
    var name: LocalizedName
    var popular: Bool
    var defaultOn: Bool
}

/// O catálogo-semente inteiro.
struct CooldownCatalog: Codable {
    var battleTasks: [BattleTask]
    var optionalTasks: [BattleTask]
    var berryTiers: [BerryTier]
    var berries: [BerryDef]

    static let empty = CooldownCatalog(battleTasks: [], optionalTasks: [], berryTiers: [], berries: [])

    /// Todas as tarefas de batalha (obrigatórias + opcionais).
    var allBattle: [BattleTask] { battleTasks + optionalTasks }

    func battleTask(_ id: String) -> BattleTask? { allBattle.first { $0.id == id } }
    func berry(_ id: String) -> BerryDef? { berries.first { $0.id == id } }
    func tier(_ id: String) -> BerryTier? { berryTiers.first { $0.tier == id } }

    static func loadFromBundle() -> CooldownCatalog {
        guard let url = Bundle.module.url(forResource: "cooldowns", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let cat = try? JSONDecoder().decode(CooldownCatalog.self, from: data)
        else { return .empty }
        return cat
    }
}

// MARK: - Estado do usuário (persistido)

/// Um "boneco" (conta/ALT) cadastrado pelo usuário.
struct GameCharacter: Codable, Identifiable, Equatable {
    var id: String
    var name: String
    /// Foto do ícone (PNG 128×128 em base64). nil = usa o monograma (1ª letra). Opcional pra
    /// não quebrar estados antigos e já viajar no blob de sync.
    var avatar: String? = nil
}

/// Progresso de uma berry plantada num canteiro. `plantedAt` é a verdade — tudo (colheita, wilt,
/// janelas de rega) deriva dele + do tier. `waterings` = quantas regas o usuário já confirmou.
struct BerryProgress: Codable, Equatable {
    var plantedAt: Double   // ms
    var waterings: Int
}

/// Estado completo do usuário. Salvo como um único blob JSON (preparado pra sync futuro).
struct CooldownState: Codable {
    var version: Int
    var characters: [GameCharacter]
    /// "charId:battleTaskId" -> ms da marcação.
    var battle: [String: Double]
    /// "charId:berryId:plot" -> progresso.
    var berry: [String: BerryProgress]
    /// Ids de tarefas de batalha que o usuário DESATIVOU (as defaultOn começam ligadas).
    var hiddenBattle: [String]
    /// Ids de berries que o usuário ATIVOU além das defaultOn.
    var enabledBerry: [String]
    var updatedAt: Double

    static let current = CooldownState(version: 2, characters: [], battle: [:], berry: [:],
                                       hiddenBattle: [], enabledBerry: [], updatedAt: 0)

    // Decoder tolerante: campos ausentes (versões antigas / campos novos) não quebram.
    enum CodingKeys: String, CodingKey {
        case version, characters, battle, berry, hiddenBattle, enabledBerry, updatedAt
    }
    init(version: Int, characters: [GameCharacter], battle: [String: Double],
         berry: [String: BerryProgress], hiddenBattle: [String], enabledBerry: [String], updatedAt: Double) {
        self.version = version; self.characters = characters; self.battle = battle
        self.berry = berry; self.hiddenBattle = hiddenBattle; self.enabledBerry = enabledBerry
        self.updatedAt = updatedAt
    }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        version = (try? c.decode(Int.self, forKey: .version)) ?? 2
        characters = (try? c.decode([GameCharacter].self, forKey: .characters)) ?? []
        battle = (try? c.decode([String: Double].self, forKey: .battle)) ?? [:]
        berry = (try? c.decode([String: BerryProgress].self, forKey: .berry)) ?? [:]
        hiddenBattle = (try? c.decode([String].self, forKey: .hiddenBattle)) ?? []
        enabledBerry = (try? c.decode([String].self, forKey: .enabledBerry)) ?? []
        updatedAt = (try? c.decode(Double.self, forKey: .updatedAt)) ?? 0
    }
}

File: windows/PrestreloAjuda/Services/CooldownModel.cs
using System.IO;
using System.Text.Json;

namespace PrestreloAjuda.Services;

// Modelo do sistema de Cooldown/Alarme (#33). Porte fiel de CooldownModel.swift do Mac.
// - CATÁLOGO: templates lidos de data/cooldowns.json (batalhas, tiers de berry, 64 berries).
// - ESTADO: o que o usuário criou/marcou (personagens + marcações), salvo em
//   %LOCALAPPDATA%\PrestreloAjuda\cooldowns.state.json (ver CooldownStore).
//
// Regra de robustez: a VERDADE é sempre um timestamp absoluto (epoch em MILISSEGUNDOS). Contador e
// alarme são derivados; reconciliar no startup a partir dos timestamps.

/// Utilidades do sistema de cooldown (funções livres do Mac: nowMs / fmtRemain / fmtHoursLabel).
public static class Cd
{
    /// Milissegundos desde 1970 (mesma base do protótipo web e do futuro sync).
    public static double NowMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    /// Formata ms restantes em "d h", "h m", "m s" ou "s" (igual ao protótipo / fmtRemain do Mac).
    /// 0 (ou negativo) → "0s"; arredonda para cima.
    public static string FmtRemain(double ms)
    {
        if (ms <= 0) return "0s";
        int s = (int)Math.Ceiling(ms / 1000.0);
        int d = s / 86400; s -= d * 86400;
        int h = s / 3600; s -= h * 3600;
        int m = s / 60; s -= m * 60;
        if (d > 0) return $"{d}d {h}h";
        if (h > 0) return $"{h}h {m:00}m";
        if (m > 0) return $"{m}m {s:00}s";
        return $"{s}s";
    }

    /// "6h", "18h", "7d" — rótulo curto de uma duração em horas (múltiplo de 24 vira dias).
    public static string FmtHoursLabel(double hours)
    {
        if (hours >= 24 && hours % 24 == 0) return $"{(int)(hours / 24)}d";
        return $"{(int)hours}h";
    }
}

// MARK: - Catálogo (data/cooldowns.json)

/// Nome bilíngue (PT/EN) vindo do catálogo.
public sealed class LocalizedName
{
    public string Pt { get; set; } = "";
    public string En { get; set; } = "";
    /// EN se o idioma atual for EN, senão PT (espelha `.localized` do Mac).
    public string Localized => Strings.Current == Lang.En ? En : Pt;
}

/// Tarefa de batalha com cooldown simples (um timer). Campos extras (type/note/source) são ignorados.
public sealed class BattleTask
{
    public string Id { get; set; } = "";
    public string Category { get; set; } = "";
    public LocalizedName Name { get; set; } = new();
    public double Hours { get; set; }
    public string Color { get; set; } = "#888888";
    public string Confidence { get; set; } = "";
    public bool DefaultOn { get; set; }
    /// Especificação do ícone: "region:x" | "trainer:x" | "item:x" | "sprite:x" | "sf:símbolo".
    /// null = cai no ponto colorido.
    public string? Icon { get; set; }
    /// Agrupador na UI (ex.: "elite4" junta as 5 ligas num submenu). null = tarefa avulsa.
    public string? Group { get; set; }
}

/// Tier de berry: a matemática do ciclo (uma vez), compartilhada por várias berries.
public sealed class BerryTier
{
    public string Tier { get; set; } = "";
    public double GrowthHours { get; set; }        // plantar -> pronta pra colher (fixo)
    public double WiltHours { get; set; }          // janela de colheita depois de pronta
    public List<double> WaterWindowsHours { get; set; } = new(); // limite (h após plantar) de cada rega
    public double WaterLeadHours { get; set; }     // disparar o lembrete tantas h ANTES do limite
    public string Yield { get; set; } = "";
    public string Id => Tier;
}

/// Uma berry específica (nome/ícone) que aponta pro seu tier.
public sealed class BerryDef
{
    public string Id { get; set; } = "";
    public string Tier { get; set; } = "";
    public LocalizedName Name { get; set; } = new();
    public bool Popular { get; set; }
    public bool DefaultOn { get; set; }
}

/// O catálogo-semente inteiro.
public sealed class CooldownCatalog
{
    public List<BattleTask> BattleTasks { get; set; } = new();
    public List<BattleTask> OptionalTasks { get; set; } = new();
    public List<BerryTier> BerryTiers { get; set; } = new();
    public List<BerryDef> Berries { get; set; } = new();

    public static readonly CooldownCatalog Empty = new();

    /// Todas as tarefas de batalha (obrigatórias + opcionais).
    public IEnumerable<BattleTask> AllBattle => BattleTasks.Concat(OptionalTasks);

    public BattleTask? FindBattleTask(string id) => AllBattle.FirstOrDefault(t => t.Id == id);
    public BerryDef? FindBerry(string id) => Berries.FirstOrDefault(b => b.Id == id);
    public BerryTier? FindTier(string id) => BerryTiers.FirstOrDefault(t => t.Tier == id);

    private static readonly JsonSerializerOptions Options = new() { PropertyNameCaseInsensitive = true };

    /// Carrega o catálogo de data/cooldowns.json ao lado do executável (copiado de ../../data pelo
    /// csproj). Mesmo padrão do SolveLoader/AppModel. Falha silenciosa → catálogo vazio.
    public static CooldownCatalog Load()
    {
        try
        {
            var path = Path.Combine(SolveLoader.DataDir, "cooldowns.json");
            if (!File.Exists(path)) return Empty;
            return JsonSerializer.Deserialize<CooldownCatalog>(File.ReadAllText(path), Options) ?? Empty;
        }
        catch { return Empty; }
    }
}

// MARK: - Estado do usuário (persistido)

/// Um "boneco" (conta/ALT) cadastrado pelo usuário.
public sealed class GameCharacter
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    /// Foto do ícone (PNG 128×128 em base64). null = usa o monograma (1ª letra).
    public string? Avatar { get; set; }
}

/// Progresso de uma berry plantada num canteiro. `PlantedAt` é a verdade — tudo (colheita, wilt,
/// janelas de rega) deriva dele + do tier. `Waterings` = quantas regas o usuário já confirmou.
public sealed class BerryProgress
{
    public double PlantedAt { get; set; }   // ms
    public int Waterings { get; set; }
}

/// Estado completo do usuário. Salvo como um único blob JSON (preparado pra sync futuro).
/// Decoder tolerante: campos ausentes mantêm os defaults (System.Text.Json só seta o que existe).
public sealed class CooldownState
{
    public int Version { get; set; } = 2;
    public List<GameCharacter> Characters { get; set; } = new();
    /// "charId:battleTaskId" -> ms da marcação.
    public Dictionary<string, double> Battle { get; set; } = new();
    /// "charId:berryId:plot" -> progresso.
    public Dictionary<string, BerryProgress> Berry { get; set; } = new();
    /// Ids de tarefas de batalha que o usuário DESATIVOU (as defaultOn começam ligadas).
    public List<string> HiddenBattle { get; set; } = new();
    /// Ids de berries que o usuário ATIVOU além das defaultOn.
    public List<string> EnabledBerry { get; set; } = new();
    public double UpdatedAt { get; set; }
}
